crates/evm/src/execute.rs

```diff
@@ -161,6 +162,152 @@ pub trait BlockExecutorProvider: Send + Sync + Clone + Unpin + 'static {
         DB: Database<Error: Into<ProviderError> + Display>;
 }
 
+/// A factory for creating block execution strategies.
+pub trait BlockExecutionStrategyFactory<DB>
+where
+    DB: Database,
+{
+    /// The specific [`BlockExecutionStrategy`] type this factory produces.
+    type Strategy: BlockExecutionStrategy<DB>;
+
+    /// The error type returned by this factory and its produced strategies.
+    type Error: From<ProviderError>
+        + From<<Self::Strategy as BlockExecutionStrategy<DB>>::Error>
+        + core::error::Error;
+
+    /// Creates a new block execution strategy instance.
+    fn create(
+        &self,
+        block: &BlockWithSenders,
+        total_difficulty: U256,
+    ) -> Result<Self::Strategy, Self::Error>;
+}
+
+/// Defines the strategy for executing a single block.
+pub trait BlockExecutionStrategy<DB> {
+    /// The error type returned by this strategy's methods.
+    type Error: From<ProviderError> + core::error::Error;
+
+    /// Configures the environment before start applying any changes.
+    fn configure_environment(&mut self) -> Result<(), Self::Error>;
+
+    /// Applies any necessary changes before executing the block's transactions.
+    fn apply_pre_execution_changes(&mut self) -> Result<(), Self::Error>;
+
+    /// Executes all transactions in the block.
+    fn execute_transactions(
+        &mut self,
+        block: &BlockWithSenders,
+    ) -> Result<(Vec<Receipt>, u64), Self::Error>;
+
+    /// Applies any necessary changes after executing the block's transactions.
+    fn apply_post_execution_changes(&mut self) -> Result<Vec<Request>, Self::Error>;
+
+    /// Returns a reference to the current state.
+    fn state_ref(&self) -> &State<DB>;
+
+    /// Sets a hook to be called after each state change during execution.
+    fn set_state_hook<F: OnStateHook + 'static>(&mut self, hook: Option<F>);
```

Comment:
from an api perspective, a with_ fn would be more convenient here, and this is where Boxed statehook would come in handy

Reply:
makes sense, done here https://github.com/paradigmxyz/reth/pull/11447/commits/003d81e1cd63dbcb47695107f92aed6fd6653991